src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralReferenceGenerator.kt

```diff
@@ -0,0 +1,37 @@
+package uk.gov.justice.digital.hmpps.hmppsinterventionsservice.service
+
+import org.springframework.stereotype.Component
+
+@Component
+class ReferralReferenceGenerator(
+  private val prefixChars: List<Char> = ('A'..'Z').toList(),
+  private val numbers: List<Char> = ('0'..'9').toList(),
+) {
+  private val ambiguousPairs = listOf(
+    Pair("O", "0"),
+    Pair("1", "I"),
+    Pair("5", "S"),
+    Pair("8", "B"),
+  )
+
+  fun generate(categoryName: String): String {
+    return generateSequence { newReference(categoryName) }
```

Comment:
kotlin level 1000 acheived

Reply:
😀 I have no idea if this is idiomatic, I tend to (sometimes unnecessarily) flock towards the functional style of things